utilities/ggd/lib/go.go

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package lib
+
+import (
+	"path/filepath"
+	"strings"
+)
+
+// ChangedPackage is a changed package.
+type ChangedPackage struct {
+	Name        string
+	OnlyDeletes bool
+}
+
+// ChangedPackages returns all changed packages.
+func ChangedPackages(changedFiles []ChangedFile, filterDirPatterns []string) []ChangedPackage {
+	changedPkgs := make(map[string]bool, len(changedFiles))
+	for _, f := range changedFiles {
+		if filterChanges(f.Filename, filterDirPatterns) {
+			continue
+		}
+		dir := filepath.Dir(f.Filename)
+		if deletes, ok := changedPkgs[dir]; ok {
+			changedPkgs[dir] = deletes && f.Deleted
+		} else {
+			changedPkgs[dir] = f.Deleted
+		}
+	}
+	pkgs := make([]ChangedPackage, 0, len(changedPkgs))
+	for p, onlyDeletes := range changedPkgs {
+		pkgs = append(pkgs, ChangedPackage{
+			Name:        p,
+			OnlyDeletes: onlyDeletes,
+		})
+	}
+	return pkgs
+}
+
+// filterChanges filters changed files based on the following heuristic,
+// only files meeting any of the following conditions affect go compilation
+// 	- any .go file
+//  - any directory called `testdata/`
+//
+// also filters any `vendor/` directory paths
+func filterChanges(input string, filterDirPatterns []string) bool {
+	for _, ptrn := range filterDirPatterns {
```

Comment:
Won't this mean we won't run tests if we just change vendor?

Reply:
It's an option passed in from the CLI, so up to the user's discretion. 

We'd use it in the following modes:
(a) CI in our internal repo which has `vendor/` checked in, we would not filter `vendor`
(b) CI in our OSS repos (where we do not checkin vendor/`), we would filter `vendor/`

Local development for both would mirror the choices made in the CI.

Reply:
Wouldn't we then be relying on local CI to verify changes to vendor? I'm kinda of the opinion that if you don't vendor you're dependencies in your repo you need to run tests in CI each time when you pull them.

Reply:
done.